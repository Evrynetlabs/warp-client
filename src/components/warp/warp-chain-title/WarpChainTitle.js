import React, { Component } from 'react'
import Card from 'react-bootstrap/Card'
import classNames from 'classnames'
import { Container, Row, Col } from 'react-bootstrap'
import PropTypes from 'prop-types'
import { ReactComponent as EvrynetIcon } from '@/assets/icons/evrynet.svg'
import { ReactComponent as StellarIcon } from '@/assets/icons/stellar.svg'
import { ReactComponent as SwapIcon } from '@/assets/icons/swap.svg'
import 'Components/warp/warp-chain-title/warpChainTitle.scss'

export default class WarpChainTitle extends Component {
  constructor(props) {
    super(props)
    const chain = {
      EVRYNET: EvrynetIcon,
      XLM: StellarIcon,
    }
    this.stylesMain = classNames({
      [this.constructor.name]: true,
    })
    this.stylesRow = `${this.stylesMain}__row`
    this.stylesAssetElem = `${this.stylesRow}__asset`
    this.stylesAssetText = `${this.stylesAssetElem}__text`
    this.stylesSwapElem = `${this.stylesRow}__swap`
    this.stylesSwapBubble = `${this.stylesSwapElem}__bubble`
    this.state = {
      styles: {
        main: this.stylesMain,
        asset: this.stylesAssetElem,
        swap: this.stylesSwapElem,
        row: this.stylesRow,
        assetText: this.stylesAssetText,
        swapBubble: this.stylesSwapBubble,
      },
      chain,
      swap: {
        src: chain.XLM,
        dest: chain.EVRYNET,
      },
    }
  }

  _updateSwapTitle(prevProps) {
    if (prevProps.isToEvrynet === this.props.isToEvrynet) return
    if (this.props.isToEvrynet) {
      this.setState({
        swap: {
          src: this.state.chain.XLM,
          dest: this.state.chain.EVRYNET,
        },
      })
      return
    }
    this.setState({
      swap: {
        src: this.state.chain.EVRYNET,
        dest: this.state.chain.XLM,
      },
    })
  }

  componentDidUpdate(prevProps) {
    this._updateSwapTitle(prevProps)
  }

  render() {
    const SrcAsset = this.state.swap.src
    const DestAsset = this.state.swap.dest
    return (
      <Card.Header className={this.state.styles.main}>
        <Container fluid>
          <Row className={this.state.styles.row}>
            <Col
              className={classNames(
                this.state.styles.asset,
                'text-format-title-light',
                'px-0',
              )}
            >
              <div className={this.state.styles.assetText}>
                <span>From</span> <SrcAsset></SrcAsset>
              </div>
            </Col>
            <Col
              className={classNames(
                this.state.styles.swap,
                'px-0',
                'flex-grow-0',
              )}
              onClick={() => {
                this.props.toggleTransferSwitch()
              }}
            >
              <div className="speech-bubble">
                <span className={this.state.styles.swapBubble}>
                  Swap account
                </span>
              </div>
              <SwapIcon></SwapIcon>
            </Col>
            <Col
              className={classNames(
                this.state.styles.asset,
                'text-format-title-light',
                'px-0',
              )}
            >
              <div className={this.state.styles.assetText}>
                <span>To</span> <DestAsset></DestAsset>
              </div>
            </Col>
          </Row>
        </Container>
      </Card.Header>
    )
  }
}

WarpChainTitle.propTypes = {
  isToEvrynet: PropTypes.bool.isRequired,
  toggleTransferSwitch: PropTypes.func.isRequired,
}
