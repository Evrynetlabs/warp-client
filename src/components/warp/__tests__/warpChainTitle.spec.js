import React from 'react'
import { shallow } from 'enzyme'
import WarpChainTitle from 'Components/warp/warp-chain-title/WarpChainTitle'

describe('WarpChainTitle', () => {
  const mock = {
    isToEvrynet: true,
    toggleTransferSwitch: jest.fn(),
  }
  let component
  beforeEach(() => {
    component = shallow(
      <WarpChainTitle
        isToEvrynet={mock.isToEvrynet}
        toggleTransferSwitch={mock.toggleTransferSwitch}
      ></WarpChainTitle>,
    )
  })
  it('should render correspondingly', () => {
    expect(component).toMatchSnapshot()
  })

  describe('When clicking toggle button', () => {
    it('should toggle a toggleTransferSwitch button', () => {
      component.find('.WarpChainTitle__swap').simulate('click')
      expect(mock.toggleTransferSwitch).toBeCalled()
    })

    it('should change source and destination of warp title', () => {
      mock.toggleTransferSwitch.mockImplementation(() => {
        component.setProps({
          isToEvrynet: false,
        })
      })
      component.find('.WarpChainTitle__swap').simulate('click')
      expect(component.state().swap).toEqual({
        dest: 'Stellar',
        src: 'Evrynet',
      })
      expect(component).toMatchSnapshot()
    })
  })
})
