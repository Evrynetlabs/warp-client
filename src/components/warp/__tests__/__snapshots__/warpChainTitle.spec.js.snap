// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarpChainTitle When clicking toggle button should change source and destination of warp title 1`] = `
<CardHeader
  className="WarpChainTitle"
>
  <Container
    fluid={true}
  >
    <ForwardRef
      className="WarpChainTitle__row"
      noGutters={false}
    >
      <Col
        className="WarpChainTitle__row__asset text-format-title-light px-0"
      >
        <div
          className="WarpChainTitle__row__asset__text"
        >
          <span>
            From
          </span>
           
          <ForwardRef(SvgEvrynet) />
        </div>
      </Col>
      <Col
        className="WarpChainTitle__row__swap px-0 flex-grow-0"
        onClick={[Function]}
      >
        <div
          className="speech-bubble"
        >
          <span
            className="WarpChainTitle__row__swap__bubble"
          >
            Swap account
          </span>
        </div>
        <ForwardRef(SvgSwap) />
      </Col>
      <Col
        className="WarpChainTitle__row__asset text-format-title-light px-0"
      >
        <div
          className="WarpChainTitle__row__asset__text"
        >
          <span>
            To
          </span>
           
          <ForwardRef(SvgStellar) />
        </div>
      </Col>
    </ForwardRef>
  </Container>
</CardHeader>
`;

exports[`WarpChainTitle should render correspondingly 1`] = `
<CardHeader
  className="WarpChainTitle"
>
  <Container
    fluid={true}
  >
    <ForwardRef
      className="WarpChainTitle__row"
      noGutters={false}
    >
      <Col
        className="WarpChainTitle__row__asset text-format-title-light px-0"
      >
        <div
          className="WarpChainTitle__row__asset__text"
        >
          <span>
            From
          </span>
           
          <ForwardRef(SvgStellar) />
        </div>
      </Col>
      <Col
        className="WarpChainTitle__row__swap px-0 flex-grow-0"
        onClick={[Function]}
      >
        <div
          className="speech-bubble"
        >
          <span
            className="WarpChainTitle__row__swap__bubble"
          >
            Swap account
          </span>
        </div>
        <ForwardRef(SvgSwap) />
      </Col>
      <Col
        className="WarpChainTitle__row__asset text-format-title-light px-0"
      >
        <div
          className="WarpChainTitle__row__asset__text"
        >
          <span>
            To
          </span>
           
          <ForwardRef(SvgEvrynet) />
        </div>
      </Col>
    </ForwardRef>
  </Container>
</CardHeader>
`;
